### PR TITLE
C#: Fix assert and assembly unloading failure when using custom script parameters with a default value 

### LINKED ISSUE #80175: C# assigning a default value to an exported custom resource property will cause error
### Godot version

v4.1.stable.mono.official

### System information

Windows 10

### Issue description

If you assign a default value to an exported custom resource property just like this:
```
[Export]
public MyResource Res { get; set; } = new();
```

Then you will get an editor error just as follows when building the solution:
```
modules/mono/glue/runtime_interop.cpp:1324 - System.ArgumentException: An item with the same key has already been added. Key: Game.MyResource
     at System.Collections.Generic.Dictionary`2.TryInsert(TKey key, TValue value, InsertionBehavior behavior)
     at System.Collections.Generic.Dictionary`2.Add(TKey key, TValue value)
     at Godot.Bridge.ScriptManagerBridge.ScriptTypeBiMap.Add(IntPtr scriptPtr, Type scriptType) in /root/godot/modules/mono/glue/GodotSharp/GodotSharp/Core/Bridge/ScriptManagerBridge.types.cs:line 23
     at Godot.Bridge.ScriptManagerBridge.AddScriptBridge(IntPtr scriptPtr, godot_string* scriptPath) in /root/godot/modules/mono/glue/GodotSharp/GodotSharp/Core/Bridge/ScriptManagerBridge.cs:line 419
```

By the way, rebuilding again without restarting the editor at this time will also cause `.NET: Failed to unload assemblies` issue.

I think the editor should at least push a clear warning instead of throwing an error like this so that I have to restart the editor even if this behavior is not allowed.

### Steps to reproduce

MyResource.cs:
```
using Godot;

namespace Game;

[GlobalClass]
public partial class MyResource : Resource
{
	public MyResource() { }
}

```

MyNode.cs:
```
using Godot;

namespace Game;

public partial class MyNode : Node
{
	[Export]
	public MyResource Res { get; set; } = new(); // can not assign a default value here
}
```

Building a project with these two cs scripts will cause editor error mentioned above.

### Minimal reproduction project

[gdcsharp-customresource-issue.zip](https://github.com/godotengine/godot/files/12242499/gdcsharp-customresource-issue.zip)

### PR DESCRIPTION
**Pull request by ipetrariu**
Fixes https://github.com/godotengine/godot/issues/80175

When an assemby reload gets triggered and a user script is loaded that uses another user script as a parameter with a default value, that script will be loaded on the spot and registered to the `ScriptManagerBridge` `ScriptTypeBiMap`. 
After the assembly reload is done, the script gets reloaded as well and an attempt is made to add it to the map, which will trigger a double insertion assert when trying to add the type to `_typeScriptMap`, but the pointer does gets added to `_scriptTypeMap` which seems to then forever prevent assembly unloading since the pointer lingers in the dictionary.

The fix is fairly naive since I'm not all that familiar with the codebase, so feel free to see this PR as more of an attempt to bring more attention to the issue in case a more comprehensive solution is necessary here (the assembly reloading error is a big workflow hamper for me as it requires a restart after any code change, or to work around the issue by not using default values which is something I'd like to avoid if possible).

MRPs can be found in the original issue - tested that reloading worked correctly without editor errors on both.

### CODE REVIEW

**Comment by akien-mga** [comment]
For the record, your commit seems not to be linked to your GitHub account. See: [Why are my commits linked to the wrong user?](https://docs.github.com/en/pull-requests/committing-changes-to-your-project/troubleshooting-commits/why-are-my-commits-linked-to-the-wrong-user) for more info.

**Comment by ipetrariu** [comment]
Thanks for the heads-up! I don't do much git/github work so I didn't realize that was the case.

File: modules/mono/csharp_script.cpp
Comment (398utubzyt) [inline modules/mono/csharp_script.cpp:987]: This can be inlined to a single if statement.
```suggestion
			String script_path = scr->get_path();

			if (GDMonoCache::godot_api_cache_updated && GDMonoCache::managed_callbacks.ScriptManagerBridge_GetExistingScriptBridgeForPath(&script_path, &existing_script)) {
```

File: modules/mono/glue/GodotSharp/GodotSharp/Core/Bridge/ScriptManagerBridge.cs
Comment (398utubzyt) [inline modules/mono/glue/GodotSharp/GodotSharp/Core/Bridge/ScriptManagerBridge.cs:462]: This should probably use `_scriptTypeBiMap.ReadWriteLock` [as seen in `GetOrCreateScriptBridgeForType`](https://github.com/godotengine/godot/pull/81541/files#diff-597dfc729c32276fb0593ca6f8bf3b2e2e847ae52217a06f60c72452a2dda3dfR470-R481).
```suggestion
                lock (_scriptTypeBiMap.ReadWriteLock)
                {
                    if (_scriptTypeBiMap.TryGetScriptPtr(scriptType, out IntPtr scriptPtr))
                    {
                        NativeFuncs.godotsharp_ref_new_from_ref_counted_ptr(out *outScript, scriptPtr);
                        return godot_bool.True;
                    }
                }
```

File: modules/mono/mono_gd/gd_mono_cache.cpp
Comment (398utubzyt) [inline modules/mono/mono_gd/gd_mono_cache.cpp:67]: Small nitpick, but these should be run in the order that the callbacks are defined.
```suggestion
	CHECK_CALLBACK_NOT_NULL(ScriptManagerBridge, GetOrCreateScriptBridgeForPath);
	CHECK_CALLBACK_NOT_NULL(ScriptManagerBridge, GetExistingScriptBridgeForPath);
```

**Review by 398utubzyt — CHANGES_REQUESTED** [review]
Not a maintainer, but here's some things I noticed while skimming through the code. Haven't tested but it looks good, otherwise!

**Review by 398utubzyt — APPROVED** [review]
Just got around to testing this! Everything seems to be working correctly and the code looks good to me. However, because the root cause of this issue is script loading order (which afaik is really hard to control, I was very rarely able to reproduce the error prior to this PR), I think it would be good to get someone else familiar with the .NET side of things to also test this before merging. Since I'm not actually a part of the Godot team, you need would someone from the .NET team to approve these changes anyway, so I'm sure that'll happen whenever they get around to it.

Also, this completely slipped my mind previously, but since you are a new contributor, make sure to read [CONTRIBUTING.md](https://github.com/godotengine/godot/blob/master/CONTRIBUTING.md) and the [contributing documentation](https://docs.godotengine.org/en/latest/contributing/workflow/pr_workflow.html) if you haven't already.

**You will also need to squash the commits before this PR can be merged**. You can find more info about squashing commits [here](https://docs.godotengine.org/en/latest/contributing/workflow/pr_workflow.html#the-interactive-rebase).

Feel free to reach out in the [development chat](https://chat.godotengine.org/) if you need help.

**Comment by ipetrariu** [comment]
Thanks for the review and resources. The PR should hopefully be set up correctly now. :)

**Comment by cjohnson57** [comment]
Please merge this @godotengine this would resolve several issues I'm having, detailed in #80175

File: modules/mono/csharp_script.cpp
Comment (paulloz) [inline modules/mono/csharp_script.cpp:986]: Wondering if those should be merged instead of replaced in this case.

File: modules/mono/csharp_script.cpp
Comment (paulloz) [inline modules/mono/csharp_script.cpp:988]: Is this necessary here? I think it'd be an issue if we end up with `CSharpScript`s in the bimap, but don't have the right path on the resource already?

**Review by paulloz — COMMENTED** [review]
Couple of questions, because I'm never sure of all the implications when dealing with the reloading logic. Overall, I think it'd be an interesting fix to get in, this issue is quite problematic for a lot of people.

File: modules/mono/csharp_script.cpp
Comment (ipetrariu) [reply modules/mono/csharp_script.cpp:988]: I believe the important part here is that we're taking over the path from the other script, since we otherwise have both of them pointing to the same path which I think was leading to asserts.

**Comment by raulsntos** [comment]
I've been unable to reproduce the issue, so I'm trying to understand what's going on here.

From what I understand reading the PR description, the issue occurs when a script is reloaded twice? It should be possible to reload a script multiple times without causing issues because when the reload is not invalidated the reload request should be ignored. If that's not happening then maybe something else is going on here.

Let's say that we have two scripts `ScriptA` and `ScriptB` defined like this:

```csharp
class ScriptA : Resource { }

class ScriptB : Node
{
    [Export]
    public ScriptA Resource { get; set } = new();
}
```

If we reload `ScriptB` before we reload `ScriptA`, then on the process of reloading `ScriptB` we call the `ScriptA` constructor which will also initialize the `CSharpScript` for `ScriptA`.

Later when `ScriptA` is reloaded, since it was already initialized before, it should not have an invalidated reload (i.e.: `reload_invalidated == false`) so reloading it does nothing.

It should not try to add the same script to the map twice, and there can't be multiple instances for the same `CSharpScript`. If there's an existing instance of the `CSharpScript` it should be retrieving it instead of creating a new one.

Can anyone who can reproduce this issue help me understand where this process fails?

**Comment by ipetrariu** [comment]
Hey @raulsntos, thanks for taking a look! I have an MRP attached to the linked issue - all you should need to do in order to reproduce the bug is to open it and build from godot or an external editor, which will lead to the following error:
`modules/mono/glue/runtime_interop.cpp:1324 - System.ArgumentException: An item with the same key has already been added. Key: resource_wrapper` 
Making any changes to any script will then lead to the assembly unloading error due to the script instance in the map essentially leaking.

It's been a while since I worked on this issue, and the order of things is a bit complicated in the first place so I hope I don't get it wrong here, but I belive the gist of it is was this: 
During a reload, if `ScriptB` is loaded first, it will lead to `ScriptA` being added immediately to the map both ways (script -> type and type -> script), but since `ScriptA` already existed previously, it will be found on the `to_reload` list during `CSharpLanguage::reload_assemblies` so during *its* reload it will try to also add itself to the map. As part of this, it will succeed in adding itself to the script->type portion of the map (since it is in essence a different instance of the script, it has a different pointer), but the type->script will see that there's already a script associated with that type, which won't ever end up getting removed from the map, as the cleanup operations don't expect multiple scripts with the same type. As for the `reload_invalidated`, as far as I can tell `ScriptA` will have that set during the `_clear()` call just before the code I modified inside `reload_assemblies`. In essence the issue is that the `ScriptA` instance created by the constructor of `ScriptB` is treated as a different script than the one that's undergoing the reload.

Hopefully that helps at least a bit, but I might be off on the details as I haven't looked into this issue in a long time (since around the time I opened this PR to be honest).

**Comment by paulloz** [comment]
What my understanding is, after a bit of digging.
1. Enter `CSharpLanguage::reload_assemblies`.
3. Build and cache a list of `CSharpScript` instance we're going to reload.
4. Everything in that list have its `reload_invalidated` set to `true`
5. Actually do the domain reload.
   1. What happens is a tad convoluted, but at some point we're going to get through the loader, and extract info about exports, etc.
   2. For some types, we path through `ScriptManagerBridge.CreateScriptBridgeForType` and create a new instance of `CSharpScript`.
   3. Those are reloaded via `CSharpScript::reload_registered_script`, setting their `reload_invalidated` to `false`.
   6. And injected into the bimap, referencing the new instances.
7. At last, we go through the list built during 2., with an already partially populated bimap. And during 3.iii., we reverted `reload_invalidated` on different instances for the same types.

**Comment by raulsntos** [comment]
Thank you both for helping me understand the issue, I've been able to reproduce it now. I think https://github.com/godotengine/godot/pull/87838 is a better solution so I'll close this PR as superseded. Thanks for the contribution nonetheless!

**Pull request closed without merging**
